lib/filewatch/observing_tail.rb

```diff
@@ -65,45 +55,35 @@ def subscribe(observer = NullObserver.new)
         end
       end # @watch.subscribe
       # when watch.subscribe ends - its because we got quit
-      _sincedb_write
+      @sincedb.write("shutting down")
     end # def subscribe
 
     private
+
     def observe_read_file(watched_file, listener)
       changed = false
-      loop do
+      @opts[:read_iterations].times do
         begin
-          data = watched_file.file_read(32768)
+          data = watched_file.file_read(FILE_READ_SIZE)
           changed = true
           watched_file.buffer_extract(data).each do |line|
             listener.accept(line)
-            @sincedb[watched_file.inode] += (line.bytesize + @delimiter_byte_size)
+            @sincedb.increment(watched_file.storage_key, line.bytesize + @delimiter_byte_size)
           end
-          # watched_file bytes_read tracks the sincedb entry
-          # see TODO in watch.rb
-          watched_file.update_bytes_read(@sincedb[watched_file.inode])
         rescue EOFError
           listener.eof
           break
         rescue Errno::EWOULDBLOCK, Errno::EINTR
           listener.error
           break
         rescue => e
-          @logger.debug? && @logger.debug("observe_read_file: general error reading #{watched_file.path} - error: #{e.inspect}")
+          @logger.error("observe_read_file: general error reading #{watched_file.path} - error: #{e.inspect}")
           listener.error
           break
         end
       end
 
-      if changed
-        now = Time.now.to_i
-        delta = now - @sincedb_last_write
-        if delta >= @opts[:sincedb_write_interval]
-          @logger.debug? && @logger.debug("writing sincedb (delta since last write = #{delta})")
-          _sincedb_write
-          @sincedb_last_write = now
-        end
-      end
+      @sincedb.write_periodically if changed
```

Comment:
I'd perhaps call this `@sincedb.request_disk_flush` or similar, what do you think?